flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java

```diff
@@ -139,119 +162,108 @@ private static ResourceCounter tryFulfillRequirementsForJobWithRegisteredResourc
-                        resultBuilder.addPendingTaskManagerAllocate(pendingTaskManager);
-                        resultBuilder.addAllocationOnPendingResource(
-                                jobId,
-                                pendingTaskManager.getPendingTaskManagerId(),
-                                effectiveProfile);
-                        availableResources.put(
-                                pendingTaskManager.getPendingTaskManagerId(),
-                                totalResourceProfile.subtract(effectiveProfile));
-                    } else {
-                        resultBuilder.addUnfulfillableJob(jobId);
-                        break;
-                    }
+                            jobId,
+                            newPendingTaskManager.getPendingTaskManagerId(),
+                            effectiveProfile);
+                    remainResource = remainResource.subtract(effectiveProfile);
                 }
+                if (!remainResource.equals(ResourceProfile.ZERO)) {
+                    availableResources.add(
+                            new InternalResourceInfo(
+                                    newPendingTaskManager.getPendingTaskManagerId(),
+                                    defaultSlotResourceProfile,
+                                    remainResource,
+                                    resultBuilder));
+                }
+            }
+        }
+    }
+
+    private static class InternalResourceInfo {
+        private final ResourceProfile defaultSlotProfile;
+        private final BiConsumer<JobID, ResourceProfile> allocationConsumer;
+        private ResourceProfile availableProfile;
+
+        InternalResourceInfo(
+                InstanceID instanceId,
+                ResourceProfile defaultSlotProfile,
+                ResourceProfile availableProfile,
+                ResourceAllocationResult.Builder resultBuilder) {
+            this.defaultSlotProfile = defaultSlotProfile;
+            this.availableProfile = availableProfile;
+            this.allocationConsumer =
+                    (jobId, slotProfile) ->
+                            resultBuilder.addAllocationOnRegisteredResource(
+                                    jobId, instanceId, slotProfile);
+        }
+
+        InternalResourceInfo(
+                PendingTaskManagerId pendingTaskManagerId,
+                ResourceProfile defaultSlotProfile,
+                ResourceProfile availableProfile,
+                ResourceAllocationResult.Builder resultBuilder) {
+            this.defaultSlotProfile = defaultSlotProfile;
+            this.availableProfile = availableProfile;
+            this.allocationConsumer =
+                    (jobId, slotProfile) ->
+                            resultBuilder.addAllocationOnPendingResource(
+                                    jobId, pendingTaskManagerId, slotProfile);
+        }
```

Comment:
It might not be necessary for `InternalResourceInfo` to understand the differences between registered and pending resources.

We could have the constructor like the following,
```
InternalResourceInfo(
                ResourceProfile defaultSlotProfile,
                ResourceProfile availableProfile,
                BiConsumer<JobID, ResourceProfile> resourceAllocationAction) {
        // xxx
}
```
and decide the resource allocation action in `getRegisteredResources` and `getPendingResources`.